### PR TITLE
PROTO Tutorial

### PR DESCRIPTION
**Pull request by DavidMansolino**
https://cyberbotics.com/doc/guide/tutorial-7-your-first-proto?version=feature-proto-tutorial

### CODE REVIEW

**Comment by DavidMansolino** [comment]
:+1:

File: docs/guide/tutorial-7-your-first-proto.md [hunk 1]
```diff
@@ -0,0 +1,90 @@
+## Tutorial 7: Your First PROTO
+
+The aim of this tutorial is to create a PROTO file with the definition of the four wheels robot from the previous tutorial.
+
+### Copy the Robot Definition
+
+For now, the definition of the robot is completely contained in the world file.
+Moving this definition in a PROTO file will allow you to use this robot without the need to completely copy its definition (either several times in the same world or in different world).
+
+> **Hands on #1**: Open the world `4_wheels_robot.wbt` created in the previous tutorial in your favorite text editor.
```
Comment (omichel) [inline docs/guide/tutorial-7-your-first-proto.md]:
```suggestion
> **Hands-on #1**: Open the world `4_wheels_robot.wbt` created in the previous tutorial in your favorite text editor.
```

File: docs/guide/tutorial-7-your-first-proto.md [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+PROTO protoName [
+]
+{
+  protoBody
+}
+```
+
+The `protoName` should be the name of the PROTO file (i.e. `FourWheelsRobot` in this case), and the `protoBody` is the definition of the root node (i.e. the [Robot](../reference/robot.md) node in this case).
+
+> **Hands on #2**: Write this default structure in your PROTO file with the correct `protoName`.
```
Comment (omichel) [inline docs/guide/tutorial-7-your-first-proto.md]:
```suggestion
> **Hands-on #2**: Write this default structure in your PROTO file with the correct `protoName`.
```

File: docs/guide/tutorial-7-your-first-proto.md [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+This new PROTO node is now available for each worlds in your current project.
+
+%figure "The PROTO is now visible in the Add a node window."
+
+![tutorial_proto.png](images/tutorial_proto.png)
+
+%end
+
+> **Hands on #3**: You can now open the `4_wheels_robot.wbt` world in Webots and add the `FourWheelsRobot` node (that you just defined).
```
Comment (omichel) [inline docs/guide/tutorial-7-your-first-proto.md]: Hands-on instructions should be imperative:
```suggestion
> **Hands-on #3**: Open the `4_wheels_robot.wbt` world in Webots and add the `FourWheelsRobot` node (that you just defined).
```

File: docs/guide/tutorial-7-your-first-proto.md [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+A second 4 wheels robot should have been added at the exact same location than the already existing one.
+
+### Adding Fields
+
+As you probably noticed, this new PROTO node doesn't have any open field, it is therefore impossible to translate, rotate or change the controller for example.
+It is very easy to add new fields to a PROTO node and to link them with internal fields.
+This should be done in the PROTO interface part (part between the `[` and the `]`).
+
+> **Hands on #4**: Edit your PROTO file in your text editor and add the definition of the `translation` and `rotation` field in the PROTO interface part:
```
Comment (omichel) [inline docs/guide/tutorial-7-your-first-proto.md]:
```suggestion
> **Hands-on #4**: Edit your PROTO file in your text editor and add the definition of the `translation` and `rotation` field in the PROTO interface part:
```

File: docs/guide/tutorial-7-your-first-proto.md [hunk 5]
```diff
@@ -0,0 +1,90 @@
+## Tutorial 7: Your First PROTO
+
+The aim of this tutorial is to create a PROTO file with the definition of the four wheels robot from the previous tutorial.
```
Comment (omichel) [inline docs/guide/tutorial-7-your-first-proto.md]:
```suggestion
The aim of this tutorial is to create a PROTO file corresponding to the four wheels robot from the previous tutorial.
```

File: docs/guide/tutorial-7-your-first-proto.md [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+### Copy the Robot Definition
+
+For now, the definition of the robot is completely contained in the world file.
+Moving this definition in a PROTO file will allow you to use this robot without the need to completely copy its definition (either several times in the same world or in different world).
```
Comment (omichel) [inline docs/guide/tutorial-7-your-first-proto.md]:
```suggestion
Moving this definition in a PROTO file will allow you to use this robot without the need to completely copy and paste its definition (either several times in the same world or in different worlds).
```

File: docs/guide/tutorial-7-your-first-proto.md [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+
+> **Hands on #1**: Open the world `4_wheels_robot.wbt` created in the previous tutorial in your favorite text editor.
+Create a new file in the `protos` folder of your project called `FourWheelsRobot.proto` and open this file in the text editor too, this file will contain the definition of your robot.
```
Comment (omichel) [inline docs/guide/tutorial-7-your-first-proto.md]:
```suggestion
Create a new empty text file in the `protos` folder of your project called `FourWheelsRobot.proto` and open this file in the text editor too, this file will contain the definition of your robot.
```

File: docs/guide/tutorial-7-your-first-proto.md [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+%figure "The PROTO is now visible in the Add a node window."
+
+![tutorial_proto.png](images/tutorial_proto.png)
+
+%end
+
+> **Hands on #3**: You can now open the `4_wheels_robot.wbt` world in Webots and add the `FourWheelsRobot` node (that you just defined).
+The node is located in `PROTO nodes (Project) / FourWheelsRobot (Robot)`.
+
+A second 4 wheels robot should have been added at the exact same location than the already existing one.
```
Comment (omichel) [inline docs/guide/tutorial-7-your-first-proto.md]:
```suggestion
A second 4 wheels robot should have been added at the exact same location as the already existing one.
```

File: docs/guide/tutorial-7-your-first-proto.md [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+
+The `protoName` should be the name of the PROTO file (i.e. `FourWheelsRobot` in this case), and the `protoBody` is the definition of the root node (i.e. the [Robot](../reference/robot.md) node in this case).
+
+> **Hands on #2**: Write this default structure in your PROTO file with the correct `protoName`.
+From the `4_wheels_robot.wbt` world file copy the robot node (starting with `Robot {` and ending with the final `}`) and paste it in your PROTO file instead of the `protoBody`.
+Finally, save the PROTO file.
+
+### Use the PROTO Node.
+
+This new PROTO node is now available for each worlds in your current project.
```
Comment (omichel) [inline docs/guide/tutorial-7-your-first-proto.md]:
```suggestion
This new PROTO node is now available for each world in your current project.
```

File: docs/guide/tutorial-7-your-first-proto.md [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+    physics Physics {
+      ...
+    }
+    controller "4_wheels_collision_avoidance"
+  }
+}
+```
+
+You can now save your simulation in Webots and revert it. The `translation` and `rotation` of the `FourWheelsRobot` node can now be changed (either in the [scene tree](the-scene-tree.md) or using the handles in the [3D view](the-3d-window.md)).
+
```
Comment (omichel) [inline docs/guide/tutorial-7-your-first-proto.md]:
```suggestion

The same mechanism could also be used to expose the `controller` field of the [Robot](../reference/robot.md) node.

```

**Comment by fabienrohrer** [comment]
Very nice. I also think it would be good to open another field which is not a root node field such as the body color. It would explain without words that it's possible to change any field.

Another minor thing: the PROTO structure could also mention the fields.

**Comment by DavidMansolino** [comment]
> Very nice. I also think it would be good to open another field which is not a root node field such as the body color. It would explain without words that it's possible to change any field.

Good idea!

> Another minor thing: the PROTO structure could also mention the fields.

Not sure to understand what you mean :-/

File: docs/guide/tutorial-7-your-first-proto.md [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+For now, the definition of the robot is completely contained in the world file.
+Moving this definition in a PROTO file will allow you to use this robot without the need to completely copy and paste its definition (either several times in the same world or in different worlds).
+
+> **Hands-on #1**: Open the world `4_wheels_robot.wbt` created in the previous tutorial in your favorite text editor.
+Create a new empty text file in the `protos` folder of your project called `FourWheelsRobot.proto` and open this file in the text editor too, this file will contain the definition of your robot.
+
+Any PROTO file should at least respect the following structure:
+```
+PROTO protoName [
```
Comment (fabienrohrer) [inline docs/guide/tutorial-7-your-first-proto.md]: > Not sure to understand what you mean :-/

Sorry :-)
Here: I would add the `protoField` label, which is an important part to understand.

**Comment by DavidMansolino** [comment]
Note: I will try to finalize this PR but in case I do not have the time (I have some other things to finish first) do not hesitate to finalize and merge it while I am on holidays.

**Comment by DavidMansolino** [comment]
test me

**Comment by DavidMansolino** [comment]
test me

**Comment by DavidMansolino** [comment]
test me

**Comment by DavidMansolino** [comment]
test me

**Comment by DavidMansolino** [comment]
test me

**Comment by DavidMansolino** [comment]
test me

**Comment by DavidMansolino** [comment]
:+1:

**Pull request merged**